runtime/Cpp/runtime/src/Lexer.h

```diff
@@ -95,9 +95,9 @@ namespace antlr4 {
     virtual void pushMode(size_t m);
     virtual size_t popMode();
 
-    template<typename T1>
-    void setTokenFactory(TokenFactory<T1> *factory)  {
-      this->_factory = factory;
+    template<typename Factory>
+    void setTokenFactory(Ref<Factory> const& factory)  {
+      this->_factory = std::static_pointer_cast<TokenFactory<CommonToken>>(factory);
```

Comment:
I'm afraid this is the wrong approach. Look at the Java source code, which uses `public void setTokenFactory(TokenFactory<?> factory)`, so it's clear the token class must be templated, not the token source class. And your cast wouldn't work if something else but a CommonTokenFactory is passed in.

This part of the C++ runtime is a bit tricky, as there's no simple equivalent for the Java generic used for it (C++ cannot directly define templated member variables). The correct approach to use here is a simple pointer. This is also what is used in the other classes (e.g. see `setInterpreter` and `addErrorListener` in Recognizer.h, which all just use a plain object pointer). The caller of `setTokenFactory` is reponsible to manage the memory used by the factory instance. This works also well with the factory function for token factories (see `TokenFactory::create`, which creates a unique pointer).

Reply:
Great, thanks for this guidance. There were a number of ways to go here; I went with one that preserved the current type of the `Lexer::_factory` member. Your feedback clarifies that changing the type of that member is ok, and I can see the rest will flow from that. I'll post a revision here shortly.